proposals/0002-monitoring.md

```diff
@@ -0,0 +1,178 @@
+* type: 
+     * counter: a monotonically increasing or decreasing numeric value (e.g. total number of requests received)
+     * gauge: a numeric value that can arbitrarily go up and down
+     * bool: a boolean value which can be `true` or `false`
+     * string: a string
+* description (optional): A human readable description of the metric
+* displayName (optional): A human readable name of the metric for display purposes if the metric name is not
+human readable. This could e.g. be the case when the metric name is a uuid.
+* tags (optional): A list of key=value pairs, which are separated by comma. 
+
+Example: 
+
+if `GET /metrics/foo` exposes:
+
+      {"fooVal": 12345} 
+
+then `OPTIONS /metrics/foo` will expose:
+
+    {"fooVal": {
+          "unit": "ms",
+          "type": "gauge",
+          "description": "The size of foo after each request",
+          "displayName": "Size of foo",
+          "tags": "app=webshop
+         }
+    } 
+
+Metadata must not change over the lifetime of a process (i.e. it is not allowed
+to return the units as seconds in one retrieval and as hours in a subsequent one).
+The reason behind it is that e.g. a monitoring agent on Kubernetes may read the
+metadata once it sees the new container and store it. It may not periodically
+re-query the process for the metadata.
+
+In fact that metadata should probably not change during the life-time of the
+whole container image or an application, as all containers spawned from it
+will be "the same" and form part of an app, where it would be confusing in
+an overall view if the same metric has different metadata.
+
+Metadata SHOULD support caching via cache control headers and SHOULD reply with a 304 Not Modified response accordingly.
+
+## Required metrics
+
+Required metrics are exposed under `/metrics/base`
+
+The following is a list of required metrics if the application uses the data. E.g. if the application does not use any data source, then there will be no data sources listed.
+
+* java.lang.* metrics from the MBeanServer (read-only metrics, no writing, no operations)
+ Q: should this really include all values?
+   * especially Garbage collector stuff is pretty convoluted
+* List of datasources with connections in use (list can be empty)
+* ...
+
+Values from the MBean server are encoded with `MBean-Name/attribute` name to retrieve a single attribute
+or only the MBean name to retrieve all (numeric) values.
+E.g. `GET /metrics/base/java.lang:type=Memory/ObjectPendingFinalizationCount` to only get that count
+or `GET /metrics/base/java.lang:type=Memory` for all attributes related to this MBean.
+
+Q: should we expose current total memory usage (heap+non heap) in a separate item? I am in favour of that as other non-JVM 
+environments do may not be able to report fine grained values, but only a total.
+
+Q: should current thread count be exposed in a separate item?
+
+
+## Vendor specific data
+
```

Comment:
This proposal is about providing metrics rather than monitoring. The monitoring should be performed by a monitoring agent on Kubernetes, which relies on the metrics data. Can this proposal be changed to be Metrics to avoid confusion?